staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go

```diff
@@ -231,31 +225,23 @@ func (o *CopyOptions) checkDestinationIsDir(dest fileSpec) error {
 			PodName:   dest.PodName,
 		},
 
-		Command:  []string{"test", "-d", dest.File},
+		Command:  []string{"test", "-d", dest.File.String()},
 		Executor: &exec.DefaultRemoteExecutor{},
 	}
 
 	return o.execute(options)
 }
 
 func (o *CopyOptions) copyToPod(src, dest fileSpec, options *exec.ExecOptions) error {
-	if len(src.File) == 0 || len(dest.File) == 0 {
-		return errFileCannotBeEmpty
-	}
-	if _, err := os.Stat(src.File); err != nil {
+	if _, err := os.Stat(src.File.String()); err != nil {
 		return fmt.Errorf("%s doesn't exist in local filesystem", src.File)
 	}
 	reader, writer := io.Pipe()
 
-	// strip trailing slash (if any)
-	if dest.File != "/" && strings.HasSuffix(string(dest.File[len(dest.File)-1]), "/") {
-		dest.File = dest.File[:len(dest.File)-1]
-	}
-
 	if err := o.checkDestinationIsDir(dest); err == nil {
 		// If no error, dest.File was found to be a directory.
 		// Copy specified src into it
-		dest.File = dest.File + "/" + path.Base(src.File)
+		dest.File = newRemotePath(dest.File.String() + "/" + src.File.Base().String())
```

Comment:
```suggestion
		dest.File = dest.File.JoinElem(src.File.Base()))
```

Reply:
Fixed.

Reply:
I actually squashed `JoinElem` and `JoinFile` into a single `Join` which requires you to pass `filePath` to make sure we clean them properly during processing instead of blindly taking a string value as a parameter.